aten/src/THC/generic/THCTensorKthValue.cu

```diff
@@ -0,0 +1,122 @@
+#ifndef THC_GENERIC_FILE
+#define THC_GENERIC_FILE "generic/THCTensorKthValue.cu"
+#else
+
+THC_API void THCTensor_(kthvalue)(THCState* state,
+                                  THCTensor* kthValue,
+                                  THCudaLongTensor* indices,
+                                  THCTensor* input,
+                                  int64_t k, int dim, int keepDim) {
+  THAssert(kthValue != NULL && indices != NULL && input != NULL);
+  THCAssertSameGPU(THCTensor_(checkGPU)(state, 3, kthValue, indices, input));
+  THArgCheck(THCTensor_(nDimension)(state, kthValue) <= MAX_CUTORCH_DIMS, 2, CUTORCH_DIM_WARNING);
+  int64_t dims = THCudaLongTensor_nDimension(state, indices);
+  THArgCheck(dims <= MAX_CUTORCH_DIMS, 3, CUTORCH_DIM_WARNING);
+  int numDims = THCTensor_(nDimension)(state, input);
+  THArgCheck(numDims <= MAX_CUTORCH_DIMS, 4, CUTORCH_DIM_WARNING);
+
+  THArgCheck(dim >= 0 && dim < numDims, 6, "dim not in range");
+
+  int64_t sliceSize = THCTensor_(size)(state, input, dim);
+  THArgCheck(k > 0 && k <= sliceSize, 5, "k not in range for dimension");
+
+  // Build the output size, which is the dim being selected set to
+  // size 1
+  THLongStorage* kthValueSize = THCTensor_(newSizeOf)(state, input);
+  THLongStorage_set(kthValueSize, dim, 1);
+  THCTensor_(resize)(state, kthValue, kthValueSize, NULL);
```

Comment:
Let's say keepdim is False. Then we should attempt to avoid the resize, because the resize changes the contiguity of the inputs, which could be bad. [kthvalue](https://pytorch.org/docs/master/torch.html?highlight=kth%20value#torch.kthvalue) supports passing in "output" tensors that the results should be stored in.

[This function](https://github.com/pytorch/pytorch/blob/b002aee0ff5eeea7172e4037b3d5936202cb6aef/aten/src/THC/THCTensor.cpp#L367-L381) might help.

Reply:
Would the same issue occur in THCTensorTopK, which does the resize as well? https://github.com/pytorch/pytorch/blob/master/aten/src/THC/generic/THCTensorTopK.cu#L27

Maybe this should be done in a follow-up PR to fix both implementations.

Reply:
True, let's leave this then.